src/libos/crates/host-socket/src/common/sockopt.rs

```diff
@@ -0,0 +1,197 @@
+use std::convert::TryFrom;
+
+use async_io::impl_ioctl_cmd;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+cfg_if::cfg_if! {
+    if #[cfg(feature = "sgx")] {
+        use libc::ocall::getsockopt as do_getsockopt;
+        use libc::ocall::setsockopt as do_setsockopt;
+    } else {
+        use libc::getsockopt as do_getsockopt;
+        use libc::setsockopt as do_setsockopt;
+    }
+}
+
+use crate::prelude::*;
+
+#[derive(Debug)]
+pub struct GetSockOptRawCmd<'a> {
+    level: i32,
+    optname: i32,
+    optval: &'a mut [u8],
+    optlen: &'a mut u32,
+}
+
+impl<'a> GetSockOptRawCmd<'a> {
+    pub fn new(level: i32, optname: i32, optval: &'a mut [u8], optlen: &'a mut u32) -> Self {
+        Self {
+            level,
+            optname,
+            optval,
+            optlen,
+        }
+    }
+
+    pub fn getsockopt(&mut self, fd: HostFd) -> Result<()> {
+        getsockopt_by_host(fd, self.level, self.optname, self.optval, self.optlen)?;
+        Ok(())
+    }
+}
+
+impl IoctlCmd for GetSockOptRawCmd<'static> {}
```

Comment:
The LibOS can only construct valid `GetSockOptRawCmd<'a>` for `'a != 'static`, but only `GetSockOptRawCmd<'static>` implements `IoctlCmd`. This means that there must be some highly unsafe code converting `GetSockOptRawCmd<'a>` to `GetSockOptRawCmd<'static>`. And yes, as expected, in a later commit, the unsafe conversion happens in `new_setsockopt_cmd` (a big unsafe function, which itself is a bad smell); it happens in a very stealth way (there is no explicit conversion). 

Converting `T<'a>` to `T<'static>` basically means workaround Rust life times and the borrow checker (then why use Rust at all?). There must be a very, very strong reason to do so. And I don't think that is our situation.

Let's discuss offline for a better solution.